lib/cucumber/step_match_search.rb

```diff
@@ -0,0 +1,67 @@
+module Cucumber
+  module StepMatchSearch
+    def self.new(search, configuration)
+      CachesStepMatch.new(
+        AssertUnambiguousMatch.new(
+          configuration.guess? ? AttemptToGuessAmbiguousMatch.new(search) : search,
+          configuration
+        )
+      )
+    end
+
+    class AssertUnambiguousMatch
+      def initialize(search, configuration)
+        @search, @configuration = search, configuration
+      end
+
+      def call(step_name)
```

Comment:
@tooky I'm wondering whether it would be better to use the test_step rather than just step_name as the protocol here. This would allow us to pull out the event notification into another one of these decorators, and it also leaves this protocol a bit more flexible. Thoughts?